File: clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

# function cloneGraph(node: Node | null, head: Node | null = null): Node | null {
#     if (!head) {
#         head = node;
#     }
    
#     if (!node) {
#         return head;
#     }
    
#     const curr = new Node();
#     curr.val = node.val;
    
#     for (let neighbor of node.neighbors) {
#         const neighborNode = cloneGraph(neighbor, head);
#         curr.neighbors.push(neighborNode);
#     }
#     // console.log(curr);
#     return curr;
# };
class Solution:
    def __init__(self):
        self.mapping = {}
        
    def cloneGraph(self, node: 'Node', head = None) -> 'Node':
        if not head:
            head = node
        
        if not node:
            self.mapping = []
            return head
        
        curr = Node(node.val)
        self.mapping[node] = curr
        for neighbor in node.neighbors:
            
            if neighbor not in self.mapping:
                neighborNode = self.cloneGraph(neighbor, head)
                curr.neighbors.append(neighborNode)
           
            if self.mapping[neighbor] not in self.mapping[node].neighbors:
                self.mapping[node].neighbors += [self.mapping[neighbor]]
                
        return curr

File: clone_graph.go
package dsa_but_make_it_classy

type Node struct {
	Val       int
	Neighbors []*Node
}

func cloneGraph(node *Node) *Node {
	if node == nil {
		return nil
	}

	mapping := make(map[int]*Node)

	clone(node, mapping)

	return mapping[node.Val]
}

// DFS through graph create a clone of the current node or pick it up from mapping if it already exists
// iterate through nodes neighbors and if we havent already made a clone that neighbor clone it otherwise
// attach the precloned object to it
func clone(node *Node, mapping map[int]*Node) *Node {
	if node == nil {
		return nil
	}

	if _, exists := mapping[node.Val]; !exists {
		mapping[node.Val] = &Node{Val: node.Val}
	}

	clonedNode := mapping[node.Val]

	for _, option := range node.Neighbors {
		if premapped, exists := mapping[option.Val]; exists {
			clonedNode.Neighbors = append(clonedNode.Neighbors, premapped)
		} else {
			clonedNeighbor := clone(option, mapping)
			clonedNode.Neighbors = append(clonedNode.Neighbors, clonedNeighbor)
		}
	}

	return clonedNode
}
